File: 07-reference-implementation/sdk/typescript/src/identity.ts
/**
 * acp/identity — Ed25519 Agent Identity + AgentID derivation (ACP-SIGN-1.0)
 *
 * An agent's cryptographic identity consists of:
 *   - An Ed25519 private/public key pair
 *   - An AgentID derived as base58(SHA-256(raw 32-byte public key))
 *   - A DID (did:key) representation
 *
 * Zero runtime dependencies — uses only Node.js built-in `node:crypto`.
 *
 * @example
 * ```typescript
 * import { AgentIdentity } from '@acp/sdk';
 *
 * const agent = AgentIdentity.generate();
 * console.log(agent.agentId);         // base58 string
 * console.log(agent.did);             // "did:key:z..."
 * console.log(agent.publicKeyBytes);  // Buffer(32)
 *
 * // Round-trip
 * const restored = AgentIdentity.fromPrivateBytes(agent.privateKeyBytes);
 * // restored.agentId === agent.agentId
 * ```
 */

import {
  generateKeyPairSync,
  createPrivateKey,
  createPublicKey,
  sign as cryptoSign,
  verify as cryptoVerify,
  createHash,
} from 'node:crypto';
import type { KeyObject } from 'node:crypto';

// ─── DER headers ──────────────────────────────────────────────────────────────
// Ed25519 PKCS8 DER prefix (16 bytes): 302e020100300506032b657004220420
const PKCS8_PREFIX = Buffer.from('302e020100300506032b657004220420', 'hex');
// Ed25519 SPKI DER prefix (12 bytes): 302a300506032b6570032100
const SPKI_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');
// Multicodec prefix for Ed25519 public key (for did:key): 0xed 0x01
const ED25519_MULTICODEC = Buffer.from([0xed, 0x01]);

// ─── Base58btc ────────────────────────────────────────────────────────────────

const BASE58_ALPHABET =
  '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

/** Base58btc encoding (Bitcoin alphabet, same as Python's _base58_encode). */
function base58Encode(data: Buffer): string {
  // Count leading zero bytes
  let leadingZeros = 0;
  for (const byte of data) {
    if (byte === 0) leadingZeros++;
    else break;
  }

  // Convert bytes → big integer
  let n = 0n;
  for (const byte of data) {
    n = n * 256n + BigInt(byte);
  }

  // Convert big integer → base58 digits (reversed)
  const digits: number[] = [];
  while (n > 0n) {
    const rem = Number(n % 58n);
    n /= 58n;
    digits.push(rem);
  }

  return (
    BASE58_ALPHABET[0].repeat(leadingZeros) +
    digits
      .reverse()
      .map((d) => BASE58_ALPHABET[d])
      .join('')
  );
}

// ─── AgentIdentity ────────────────────────────────────────────────────────────

/**
 * Ed25519 Agent Identity (ACP-SIGN-1.0).
 *
 * Attributes:
 *   agentId        — base58(SHA-256(raw 32-byte public key))  [ACP-CT-1.0 §3]
 *   did            — did:key:z<base58(0xed01 || pubkey)>
 *   publicKeyBytes — 32-byte raw Ed25519 public key
 *   privateKeyBytes— 32-byte raw Ed25519 private key (store securely!)
 */
export class AgentIdentity {
  private readonly _privateKey: KeyObject;
  private readonly _publicKey: KeyObject;
  private readonly _pubkeyRaw: Buffer;

  private constructor(privateKey: KeyObject, publicKey: KeyObject) {
    this._privateKey = privateKey;
    this._publicKey = publicKey;
    // Strip the 12-byte SPKI DER header to get the raw 32-byte key
    const spkiDer = this._publicKey.export({ type: 'spki', format: 'der' }) as Buffer;
    this._pubkeyRaw = Buffer.from(spkiDer.subarray(12));
  }

  // ─── Constructors ───────────────────────────────────────────────────────────

  /** Generate a new random Ed25519 agent identity. */
  static generate(): AgentIdentity {
    const { privateKey, publicKey } = generateKeyPairSync('ed25519');
    return new AgentIdentity(privateKey, publicKey);
  }

  /**
   * Restore an identity from 32 raw private key bytes.
   * @param data 32-byte raw Ed25519 private key seed
   */
  static fromPrivateBytes(data: Buffer): AgentIdentity {
    if (data.length !== 32) {
      throw new Error(
        `AgentIdentity.fromPrivateBytes: expected 32 bytes, got ${data.length}`
      );
    }
    const pkcs8 = Buffer.concat([PKCS8_PREFIX, data]);
    const privateKey = createPrivateKey({ key: pkcs8, format: 'der', type: 'pkcs8' });
    const publicKey = createPublicKey(privateKey);
    return new AgentIdentity(privateKey, publicKey);
  }

  // ─── Properties ─────────────────────────────────────────────────────────────

  /**
   * ACP AgentID: base58(SHA-256(raw 32-byte public key)).
   * Implements ACP-SIGN-1.0 §3.1.
   */
  get agentId(): string {
    const digest = createHash('sha256').update(this._pubkeyRaw).digest();
    return base58Encode(digest);
  }

  /**
   * did:key representation (multicodec Ed25519 + base58btc).
   * Format: did:key:z<base58(0xed01 || pubkey)>
   */
  get did(): string {
    const multicodec = Buffer.concat([ED25519_MULTICODEC, this._pubkeyRaw]);
    return `did:key:z${base58Encode(multicodec)}`;
  }

  /** Raw 32-byte Ed25519 public key. */
  get publicKeyBytes(): Buffer {
    return Buffer.from(this._pubkeyRaw);
  }

  /** Raw 32-byte Ed25519 private key (store securely!). */
  get privateKeyBytes(): Buffer {
    const pkcs8 = this._privateKey.export({ type: 'pkcs8', format: 'der' }) as Buffer;
    return Buffer.from(pkcs8.subarray(16)); // strip 16-byte PKCS8 header
  }

  // ─── Signing ────────────────────────────────────────────────────────────────

  /**
   * Sign arbitrary bytes with Ed25519. Returns 64-byte raw signature.
   * The caller is responsible for any pre-hashing (e.g. SHA-256 for ACP-SIGN-1.0).
   */
  sign(message: Buffer): Buffer {
    return cryptoSign(null, message, this._privateKey) as Buffer;
  }

  /**
   * Verify a 64-byte Ed25519 signature against a message.
   * Returns true if valid, false otherwise.
   */
  verify(signature: Buffer, message: Buffer): boolean {
    try {
      return cryptoVerify(null, message, this._publicKey, signature);
    } catch {
      return false;
    }
  }

  toString(): string {
    return `AgentIdentity(agentId=${this.agentId})`;
  }
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

/**
 * Derive an AgentID from raw 32-byte Ed25519 public key bytes.
 * AgentID = base58(SHA-256(publicKeyBytes))
 */
export function deriveAgentId(publicKeyBytes: Buffer): string {
  const digest = createHash('sha256').update(publicKeyBytes).digest();
  return base58Encode(digest);
}

File: 07-reference-implementation/sdk/python/acp/identity.py
"""
acp.identity — Ed25519 Agent Identity + AgentID derivation (ACP-SIGN-1.0)

An agent's cryptographic identity consists of:
  - An Ed25519 private/public key pair
  - An AgentID derived as the SHA-256 of the canonical DER public key bytes,
    encoded as base64url without padding

Usage:
    from acp.identity import AgentIdentity

    # Generate a new random identity
    agent = AgentIdentity.generate()
    print(agent.agent_id)   # "acp:agent:<base64url>"
    print(agent.did)        # "did:key:z<base58btc-encoded>"

    # Load from existing private key bytes
    agent = AgentIdentity.from_private_bytes(private_key_bytes)

    # Export private key for storage
    private_bytes = agent.private_key_bytes  # 32 raw bytes
"""
from __future__ import annotations

import base64
import hashlib
from typing import Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.serialization import (
    Encoding,
    PublicFormat,
    PrivateFormat,
    NoEncryption,
)

# multicodec prefix for Ed25519 public key: 0xed 0x01
_ED25519_MULTICODEC = bytes([0xED, 0x01])


def _base64url(data: bytes) -> str:
    """Base64url encoding without padding."""
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def _base58_encode(data: bytes) -> str:
    """Base58btc encoding (Bitcoin alphabet)."""
    alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
    n = int.from_bytes(data, "big")
    result = []
    while n:
        n, remainder = divmod(n, 58)
        result.append(alphabet[remainder])
    # Leading zeros
    for byte in data:
        if byte == 0:
            result.append(alphabet[0])
        else:
            break
    return "".join(reversed(result))


class AgentIdentity:
    """
    Ed25519 Agent Identity (ACP-SIGN-1.0).

    Attributes:
        agent_id: Canonical ACP agent identifier ("acp:agent:<base64url>")
        did:      did:key representation of the public key
        public_key_bytes: 32-byte raw Ed25519 public key
    """

    def __init__(self, private_key: Ed25519PrivateKey) -> None:
        self._private_key = private_key
        self._public_key: Ed25519PublicKey = private_key.public_key()
        # Raw 32-byte public key
        self._pubkey_raw: bytes = self._public_key.public_bytes(
            Encoding.Raw, PublicFormat.Raw
        )

    # ─── Constructors ──────────────────────────────────────────────────────────

    @classmethod
    def generate(cls) -> "AgentIdentity":
        """Generate a new random Ed25519 agent identity."""
        return cls(Ed25519PrivateKey.generate())

    @classmethod
    def from_private_bytes(cls, data: bytes) -> "AgentIdentity":
        """Load from 32-byte raw private key."""
        return cls(Ed25519PrivateKey.from_private_bytes(data))

    # ─── Properties ───────────────────────────────────────────────────────────

    @property
    def agent_id(self) -> str:
        """ACP AgentID: sha-256 of raw public key, base64url-encoded."""
        digest = hashlib.sha256(self._pubkey_raw).digest()
        return f"acp:agent:{_base64url(digest)}"

    @property
    def did(self) -> str:
        """did:key representation (multicodec Ed25519 + base58btc)."""
        multicodec_bytes = _ED25519_MULTICODEC + self._pubkey_raw
        return f"did:key:z{_base58_encode(multicodec_bytes)}"

    @property
    def public_key_bytes(self) -> bytes:
        """Raw 32-byte Ed25519 public key."""
        return self._pubkey_raw

    @property
    def private_key_bytes(self) -> bytes:
        """Raw 32-byte Ed25519 private key (store securely!)."""
        return self._private_key.private_bytes(
            Encoding.Raw, PrivateFormat.Raw, NoEncryption()
        )

    # ─── Signing ──────────────────────────────────────────────────────────────

    def sign(self, message: bytes) -> bytes:
        """Sign arbitrary bytes. Returns 64-byte Ed25519 signature."""
        return self._private_key.sign(message)

    def verify(self, signature: bytes, message: bytes) -> bool:
        """Verify a signature. Returns True if valid, False otherwise."""
        try:
            self._public_key.verify(signature, message)
            return True
        except Exception:
            return False

    def __repr__(self) -> str:
        return f"AgentIdentity(agent_id={self.agent_id!r})"
